### PR TITLE
Encapsulate isogeny mapping in a new struct

### PR DESCRIPTION
**Pull request by drskalman**
## Description
This PR moves the definition of the coefficients of the isogeny map of WB hashing to a separate struct rather than defining it in the WBParams itself. This is done to declutter the definition of the curve and make it explicit that the four polynomials are related and defining a single map  according to [these suggestions from @mmagician 
 and @Pratyush](https://github.com/w3f/arkworks-curves/pull/11#discussion_r992418385)

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to GitHub issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the GitHub PR explorer

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 ## Pending
 
+- [\#517](https://github.com/arkworks-rs/algebra/pull/517) (`ark-ec`) Move the definition of the isogeny map of WB hash-to-curve to a separate struct
```
Comment (Pratyush) [inline CHANGELOG.md:5]: This should go in `Breaking changes`, I believe.

File: ec/src/hashing/curve_maps/wb/mod.rs [hunk 2]
```diff
@@ -13,6 +13,25 @@ use crate::{
...
+/// RationalMap is used to define a rational map between curves where
+/// x coordinate of the codomain poin only depends on x coordinate of domain point
+/// and the y coordinate of the codomain point is a multiple of domain point's y coordinate
+/// where the mulitplier depends on x coordinate of the domain point
+/// Phi := (a(x), b(x)*y)
+/// All isogeny maps of curves of short Weierstrass form can be written in this way. See
+/// [\[Ga18]\]. Theorem 9.7.5.
+///
+/// - [\[Ga18]\] Galbraith, S. D. (2018). Mathematics of public key cryptography. :
+///  Cambridge University Press.
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/wb/mod.rs:25]:
```suggestion
/// [`RationalMap`] is defines a rational map between curves where
/// the `x` coordinate of the codomain point only depends on the 
/// `x`-coordinate of the domain point, and the
/// `y`-coordinate of the codomain point is a multiple of the `y`-coordinate of the domain point. 
/// The multiplier depends on the `x`-coordinate of the domain point.
/// That is, the map is of the form `Phi(x, y) := (a(x), b(x)*y)`.
/// All isogeny maps of curves of short Weierstrass form can be written in this way. See
/// [\[Ga18]\]. Theorem 9.7.5 for details.
///
/// - [\[Ga18]\] Galbraith, S. D. (2018). Mathematics of public key cryptography.
```

File: ec/src/hashing/curve_maps/wb/mod.rs [hunk 3]
```diff
@@ -24,22 +43,18 @@ pub trait WBParams: SWCurveConfig + Sized {
...
     type IsogenousCurve: SWUParams<BaseField = BaseField<Self>>;
 
-    const PHI_X_NOM: &'static [BaseField<Self>];
-    const PHI_X_DEN: &'static [BaseField<Self>];
-
-    const PHI_Y_NOM: &'static [BaseField<Self>];
-    const PHI_Y_DEN: &'static [BaseField<Self>];
+    const PHI: RationalMap<'static, Self>;
 
     fn isogeny_map(
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/wb/mod.rs:48]: It would be cleaner to make this a method on `RationalMap`, I think?

**Review by Pratyush — CHANGES_REQUESTED** [review]
This looks great to me, modulo some small changes.

File: ec/src/hashing/curve_maps/wb/mod.rs [hunk 4]
```diff
@@ -13,45 +13,69 @@ use crate::{
...
-
-    const PHI_X_NOM: &'static [BaseField<Self>];
-    const PHI_X_DEN: &'static [BaseField<Self>];
-
-    const PHI_Y_NOM: &'static [BaseField<Self>];
-    const PHI_Y_DEN: &'static [BaseField<Self>];
+/// - [\[Ga18]\] Galbraith, S. D. (2018). Mathematics of public key cryptography.
+pub struct IsogenyMap<
+    'a,
+    DOMAINE: SWCurveConfig,
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/wb/mod.rs:28]:
```suggestion
    DOMAIN: SWCurveConfig,
```

File: ec/src/hashing/curve_maps/wb/mod.rs [hunk 5]
```diff
@@ -13,45 +13,69 @@ use crate::{
...
+    CODOMAINE: SWCurveConfig<BaseField = BaseField<DOMAINE>>,
+> {
+    pub x_map_numerator: &'a [BaseField<CODOMAINE>],
+    pub x_map_denominator: &'a [BaseField<CODOMAINE>],
+
+    pub y_map_numerator: &'a [BaseField<CODOMAINE>],
+    pub y_map_denominator: &'a [BaseField<CODOMAINE>],
+
+    pub _phantom_domain: PhantomData<DOMAINE>,
+}
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/wb/mod.rs:38]:
```suggestion
pub struct IsogenyMap<
    'a,
    DOMAIN: SWCurveConfig,
    CODOMAIN: SWCurveConfig<BaseField = BaseField<DOMAIN>>,
> {
    pub x_map_numerator: &'a [BaseField<CODOMAIN>],
    pub x_map_denominator: &'a [BaseField<CODOMAIN>],

    pub y_map_numerator: &'a [BaseField<CODOMAIN>],
    pub y_map_denominator: &'a [BaseField<CODOMAIN>],
}
```

File: ec/src/hashing/curve_maps/wb/mod.rs [hunk 6]
```diff
@@ -13,45 +13,69 @@ use crate::{
...
 
-    fn isogeny_map(
-        domain_point: Affine<Self::IsogenousCurve>,
-    ) -> Result<Affine<Self>, HashToCurveError> {
+impl<'a, DOMAINE, CODOMAINE> IsogenyMap<'a, DOMAINE, CODOMAINE>
+where
+    DOMAINE: SWCurveConfig,
+    CODOMAINE: SWCurveConfig<BaseField = BaseField<DOMAINE>>,
+{
+    fn apply(&self, domain_point: Affine<DOMAINE>) -> Result<Affine<CODOMAINE>, HashToCurveError> {
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/wb/mod.rs:45]:
```suggestion
impl<'a, DOMAIN, CODOMAIN> IsogenyMap<'a, DOMAIN, CODOMAIN>
where
    DOMAIN: SWCurveConfig,
    CODOMAIN: SWCurveConfig<BaseField = BaseField<DOMAIN>>,
{
    fn apply(&self, domain_point: Affine<DOMAIN>) -> Result<Affine<CODOMAIN>, HashToCurveError> {
```

File: ec/src/hashing/curve_maps/wb/mod.rs [hunk 7]
```diff
@@ -13,45 +13,69 @@ use crate::{
...
+                let x_num = DensePolynomial::from_coefficients_slice(self.x_map_numerator);
+                let x_den = DensePolynomial::from_coefficients_slice(self.x_map_denominator);
 
-                let y_num = DensePolynomial::from_coefficients_slice(Self::PHI_Y_NOM);
-                let y_den = DensePolynomial::from_coefficients_slice(Self::PHI_Y_DEN);
+                let y_num = DensePolynomial::from_coefficients_slice(self.y_map_numerator);
+                let y_den = DensePolynomial::from_coefficients_slice(self.y_map_denominator);
 
-                let mut v: [BaseField<Self>; 2] = [x_den.evaluate(x), y_den.evaluate(x)];
+                let mut v: [BaseField<DOMAINE>; 2] = [x_den.evaluate(x), y_den.evaluate(x)];
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/wb/mod.rs:54]:
```suggestion
                let mut v: [BaseField<DOMAIN>; 2] = [x_den.evaluate(x), y_den.evaluate(x)];
```

File: ec/src/hashing/curve_maps/wb/mod.rs [hunk 8]
```diff
@@ -13,45 +13,69 @@ use crate::{
...
                 batch_inversion(&mut v);
                 let img_x = x_num.evaluate(x) * v[0];
                 let img_y = (y_num.evaluate(x) * y) * v[1];
-                Ok(Affine::new_unchecked(img_x, img_y))
+                Ok(Affine::<CODOMAINE>::new_unchecked(img_x, img_y))
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/wb/mod.rs:58]:
```suggestion
                Ok(Affine::<CODOMAIN>::new_unchecked(img_x, img_y))
```

File: ec/src/hashing/curve_maps/wb/mod.rs [hunk 9]
```diff
@@ -273,7 +300,15 @@ mod test {
             MontFp!("-63"),
             MontFp!("44"),
             MontFp!("1"),
-        ];
+        ],
+
+        _phantom_domain: PhantomData::<TestSWU127MapToIsogenousCurveParams>,
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/wb/mod.rs:305]:
```suggestion
```

File: test-curves/src/bls12_381/g1_swu_iso.rs [hunk 10]
```diff
@@ -58,6 +60,77 @@ impl SWUParams for SwuIsoParameters {
     const ZETA: Fq = MontFp!("11");
 }
 
+pub const ISOGENY_MAP_TO_G1 : IsogenyMap<'_, SwuIsoParameters, g1::Parameters, > = IsogenyMap::<'static, SwuIsoParameters, g1::Parameters> {
```
Comment (Pratyush) [inline test-curves/src/bls12_381/g1_swu_iso.rs:63]:
```suggestion
pub const ISOGENY_MAP_TO_G1 : IsogenyMap<'_, SwuIsoParameters, g1::Parameters, > = IsogenyMap {
```

File: test-curves/src/bls12_381/g1_swu_iso.rs [hunk 11]
```diff
@@ -58,6 +60,77 @@ impl SWUParams for SwuIsoParameters {
...
+        MontFp!("3459661102222301807083870307127272890283709299202626530836335779816726101522661683404130556379097384249447658110805"),
+        MontFp!("742483168411032072323733249644347333168432665415341249073150659015707795549260947228694495111018381111866512337576"),
+        MontFp!("1662231279858095762833829698537304807741442669992646287950513237989158777254081548205552083108208170765474149568658"),
+        MontFp!("1668238650112823419388205992952852912407572045257706138925379268508860023191233729074751042562151098884528280913356"),
+        MontFp!("369162719928976119195087327055926326601627748362769544198813069133429557026740823593067700396825489145575282378487"),
+        MontFp!("2164195715141237148945939585099633032390257748382945597506236650132835917087090097395995817229686247227784224263055"),
+        MontFp!("1"),
+    ],
+
+    _phantom_domain : PhantomData::<SwuIsoParameters>,
```
Comment (Pratyush) [inline test-curves/src/bls12_381/g1_swu_iso.rs:131]:
```suggestion
```

File: test-curves/src/bls12_381/g2_swu_iso.rs [hunk 12]
```diff
@@ -75,6 +76,68 @@ impl SWUParams for SwuIsoParameters {
     // ZETA = -(2 + u) as per IETF draft.
     const ZETA: Fq2 = Fq2::new(MontFp!("-2"), MontFp!("-1"));
 }
+
+pub const ISOGENY_MAP_TO_G2  : IsogenyMap<'_, SwuIsoParameters, g2::Parameters> = IsogenyMap::<'static, SwuIsoParameters, g2::Parameters> {
```
Comment (Pratyush) [inline test-curves/src/bls12_381/g2_swu_iso.rs:80]:
```suggestion
pub const ISOGENY_MAP_TO_G2  : IsogenyMap<'_, SwuIsoParameters, g2::Parameters> = IsogenyMap {
```

File: test-curves/src/bls12_381/g2_swu_iso.rs [hunk 13]
```diff
@@ -75,6 +76,68 @@ impl SWUParams for SwuIsoParameters {
...
+                   MontFp!("4002409555221667393417789825735904156556882819939007885332058136124031650490837864442687629129015664037894272559571")),
+        Fq2::new(
+                   MontFp!("18"),
+                   MontFp!("4002409555221667393417789825735904156556882819939007885332058136124031650490837864442687629129015664037894272559769")),
+        Fq2::new(
+                   MontFp!("1"),
+                   MontFp!("0")),
+    ],
+
+    _phantom_domain : PhantomData::<SwuIsoParameters>,
```
Comment (Pratyush) [inline test-curves/src/bls12_381/g2_swu_iso.rs:138]:
```suggestion
```

File: ec/src/hashing/curve_maps/wb/mod.rs [hunk 14]
```diff
@@ -195,10 +220,12 @@ mod test {
...
-    impl WBParams for TestWBF127MapToCurveParams {
-        type IsogenousCurve = TestSWU127MapToIsogenousCurveParams;
-
-        const PHI_X_NOM: &'static [<Self::IsogenousCurve as CurveConfig>::BaseField] = &[
+    const ISOGENY_MAP_TESTWBF127: IsogenyMap<
+        '_,
+        TestSWU127MapToIsogenousCurveParams,
+        TestWBF127MapToCurveParams,
+    > = IsogenyMap::<'static, TestSWU127MapToIsogenousCurveParams, TestWBF127MapToCurveParams> {
+        x_map_numerator: &[
```
Comment (Pratyush) [inline ec/src/hashing/curve_maps/wb/mod.rs:228]:
```suggestion
    > = IsogenyMap {
        x_map_numerator: &[
```

**Review by Pratyush — APPROVED** [review]

File: ec/src/hashing/curve_maps/wb/mod.rs (hunk 5 above)
Comment (Pratyush) [reply ec/src/hashing/curve_maps/wb/mod.rs]: Ah, sorry, didn't realize that we we weren't using `DOMAIN`. There's two options:
* Reinstate the `PhantomData<CODOMAIN>` field, but make it private. To enable constructing an `IsogenyMap`, let's add a `new` method which does the construction.
* Change the `*_map_denominator` type signatures to be `&'a [BaseField<DOMAIN>]`. Admittedly this is a bit of a hack, but this avoids adding back the field.

Comment (drskalman) [reply ec/src/hashing/curve_maps/wb/mod.rs]: > * To enable constructing an `IsogenyMap`, let's add a `new` method which does the construction.

Hmm, then new need to be a const function

Comment (Pratyush) [reply ec/src/hashing/curve_maps/wb/mod.rs]: That should be possible, since it's a simple method, no?

Comment (drskalman) [reply ec/src/hashing/curve_maps/wb/mod.rs]: Maybe I'll go with the second solution then.

Comment (Pratyush) [reply ec/src/hashing/curve_maps/wb/mod.rs]: We only need the method to be a `const fn`; the arguments don't have to be explicitly `const`; by calling `new` in a const context, the arguments are constrained to be `const`

Comment (drskalman) [reply ec/src/hashing/curve_maps/wb/mod.rs]: hmm but they need to have static lifetime no?

Comment (drskalman) [reply ec/src/hashing/curve_maps/wb/mod.rs]: I try sending them as array instead of refrences...

Comment (drskalman) [reply ec/src/hashing/curve_maps/wb/mod.rs]: I can't because we don't know they don't have the same size across the curves.

Comment (drskalman) [reply ec/src/hashing/curve_maps/wb/mod.rs]: @Pratyush [here](https://github.com/arkworks-rs/algebra/pull/517) is the code with new constructor. If you tell me how to send the slices with static lifetime to new function so it compiles, then I'll update this PR accordingly.

**Pull request merged**